netperf_tool/src/server_node.cpp

```diff
@@ -0,0 +1,225 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "server_node.hpp"
+
+#include <chrono>
+#include <condition_variable>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "rclcpp/logging.hpp"
+
+#include "utils.hpp"
+
+namespace netperf_tool
+{
+
+ServerNode::ServerNode(
+  const rclcpp::NodeOptions & options,
+  const rclcpp::QoS & sub_qos)
+: Node("server", "netperf_tool", options)
+{
+  sub_ = this->create_subscription<netperf_tool_interfaces::msg::Bytes>(
+    "test_topic", sub_qos,
+    std::bind(&ServerNode::handle_msg, this, std::placeholders::_1, std::placeholders::_2));
+  srv_ = this->create_service<netperf_tool_interfaces::srv::GetResults>(
+    "get_results",
+    std::bind(
+      &ServerNode::handle_get_results_request,
+      this,
+      std::placeholders::_1,
+      std::placeholders::_2));
+}
+
+void
+ServerNode::handle_msg(
+  const rclcpp::SerializedMessage & serialized_msg,
+  const rclcpp::MessageInfo & msg_info)
+{
+  auto rec_timestamp = std::chrono::system_clock::now();
+  std::chrono::nanoseconds latency;
+  size_t received_bytes = serialized_msg.size();
+  netperf_tool_interfaces::msg::Bytes msg;
+  deserializer_.deserialize_message(&serialized_msg, &msg);
+  const auto & rmw_msg_info = msg_info.get_rmw_message_info();
+  if (rmw_msg_info.source_timestamp != 0 && rmw_msg_info.received_timestamp != 0) {
+    // source and received timestamp supported
+    // TODO(ivanpauno): Add a way to evaluate this in rmw_feature_supported() to avoid the if().
+    RCLCPP_INFO_ONCE(
+      this->get_logger(),
+      "source and received timestamp supported, using them to calculate latency!");
+    auto pub_timestamp = std::chrono::time_point<std::chrono::system_clock>{
+      std::chrono::nanoseconds{rmw_msg_info.source_timestamp}};
+    rec_timestamp =
+      std::chrono::time_point<std::chrono::system_clock>{
+      std::chrono::nanoseconds{rmw_msg_info.received_timestamp}};
+    latency = rec_timestamp - pub_timestamp;
+  } else {
+    // TODO(ivanpauno): Maybe it's worth to always calculate this latency (?)
+    // as it takes in account the executor overhead.
+    RCLCPP_INFO_ONCE(
+      this->get_logger(),
+      "either source or received timestamp are not supported,"
+      "using timestamps added to the message to calculate latency");
+    auto pub_timestamp =
+      std::chrono::time_point<std::chrono::system_clock>{
+      std::chrono::nanoseconds{msg.timestamp}};
```

Comment:
This could become:

```c++
    auto pub_timestamp =
      std::chrono::time_point<std::chrono::system_clock>{
      std::chrono::nanoseconds{rclcpp::Time(msg.timestamp).nanoseconds}};
```

Reply:
Done in https://github.com/ros2/ros2cli/pull/720/commits/084b304892a29231a93a5a0f46fa8b9ad349b639